codex/slots/proofs/backends/circomcompat.nim

```diff
@@ -0,0 +1,237 @@
+## Nim-Codex
+## Copyright (c) 2024 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+{.push raises: [].}
+
+import std/sequtils
+
+import pkg/chronos
+import pkg/questionable/results
+import pkg/circomcompat
+import pkg/poseidon2/io
+
+import ../../types
+import ../../../stores
+import ../../../merkletree
+
+import pkg/constantine/math/arithmetic
+import pkg/constantine/math/arithmetic
+import pkg/constantine/math/io/io_bigints
+
+export circomcompat
+
+const
+  # TODO: this defaults need to be adjusted and/or replased with cli config params
+  DefaultMaxSlotDepth*    = 32
+  DefaultMaxDatasetDepth* = 8
+  DefaultBlockDepth*      = 5
+  DefaultCellElms*        = 67
+  DefaultSamplesNum*      = 5
+
+type
+  CircomCompat* = object
+    slotDepth     : int     # max depth of the slot tree
+    datasetDepth  : int     # max depth of dataset  tree
+    blkDepth      : int     # depth of the block merkle tree (pow2 for now)
+    cellElms      : int     # number of field elements per cell
+    numSamples    : int     # number of samples per slot
+    r1csPath      : string  # path to the r1cs file
+    wasmPath      : string  # path to the wasm file
+    zKeyPath      : string  # path to the zkey file
+    backendCfg    : ptr CircomBn254Cfg
+
+  CircomG1* = G1
+  CircomG2* = G2
+
+  CircomProof*  = Proof
+  CircomKey*    = VerifyingKey
+  CircomInputs* = Inputs
+
+proc release*(self: CircomCompat) =
+  ## Release the backend
+  ##
+
+  self.backendCfg.unsafeAddr.releaseCfg()
+
+proc getVerifyingKey*(
+  self: CircomCompat): ?!ptr CircomKey =
+  ## Get the verifying key
+  ##
+
+  var
+    cfg: ptr CircomBn254Cfg = self.backendCfg
+    vkpPtr: ptr VerifyingKey = nil
+
+  if cfg.getVerifyingKey(vkpPtr.addr) != ERR_OK or vkpPtr == nil:
+    return failure("Failed to get verifying key")
+
+  success vkpPtr
+
+proc prove*[H](
+  self: CircomCompat,
+  input: ProofInput[H]): ?!CircomProof =
+  ## Encode buffers using a backend
+  ##
+
+  # NOTE: All inputs are statically sized per circuit
+  # and adjusted accordingly right before being passed
+  # to the circom ffi - `setLen` is used to adjust the
+  # sequence length to the correct size which also 0 pads
+  # to the correct length
+  doAssert input.samples.len == self.numSamples,
+    "Number of samples does not match"
+
+  doAssert input.slotProof.len <= self.datasetDepth,
+    "Number of slot proofs does not match"
+
+  doAssert input.samples.allIt(
+    block:
+      (it.merklePaths.len <= self.slotDepth + self.blkDepth and
```

Comment:
I'm guessing blkDepth is the subpath that goes into the 64k block, correct?

Reply:
exactly